runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -118,14 +120,16 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
             try {
               for (;;) {
                 Optional<Boolean> result = checkForSuccess(job);
-                if (result.isPresent()) {
+                if (result.isPresent() && (!result.get() || atMaxWatermark(job))) {
                   return result;
                 }
                 Thread.sleep(10000L);
               }
             } finally {
-              LOG.info("Cancelling Dataflow job {}", job.getJobId());
-              job.cancel();
+              if (!job.getState().isTerminal()) {
```

Comment:
Oh hm, reading. If this conditional is false the job is in a terminal state and doesn't need to be cancelled. Looks good.